Documentation: document breaking change CS8153 when going to 4.0

Tests using pre-4.0 NSubstitute `Arg` matchers will start to fail if a matcher is used in an expression tree. We should document this in breaking changes.

For example:

```
        protected void specify(Expression<Action> expectation)
        {
            // Real implementation aggregates multiple specifications into a single test failure.
            // For repro can just invoke the expression directly:
            expectation.Compile().Invoke();
        }

        public interface IFoo { void DoStuff(int i); }

        [Test]
        public void Sample() {
            var sub = Substitute.For<IFoo>();
            sub.DoStuff(42);
            specify(() => sub.Received().DoStuff(Arg.Any<int>()));
        }
```

After 4.0 this fails with:

```
error CS8153: An expression tree lambda may not contain a call to a method, property, or indexer that returns by reference
```

--- a/BreakingChanges.md
+++ b/BreakingChanges.md
@@ -17,7 +17,7 @@ The argument matcher change to support `out` and `ref` mentioned above can also 
 
 > CS8153: An expression tree lambda may not contain a call to a method, property, or indexer that returns by reference
 
-Reason: `out` and `ref` support for argument matchers is important to address as it often caused confusion for people now that C# has more support for `out` and `ref`. We are always reluctant to introducing breaking changes, but in this case we estimated the benefit of the feature would outweigh the downside of no longer being able to use matchers in expression trees.
+Reason: Previous NSubstitute releases had very limited support for matching `out` and `ref` arguments, and this frequently caused confusion for people. We are always reluctant to introducing breaking changes, but in this case we estimated the benefit of improving support for matching `out` and `ref` arguments would outweigh the downside of no longer being able to use matchers in expression trees.
 
 Workaround:
 * Move the NSubstitute statement/assertion outside of the expression tree; or

--- a/BreakingChanges.md
+++ b/BreakingChanges.md
@@ -1,13 +1,39 @@
 4.0.0 Release
 ================
 
-Argument matchers (`Arg.Is`, `Arg.Any` etc.) now use [`ref` returns which were introduced in C# 7.0](https://blogs.msdn.microsoft.com/dotnet/2016/08/24/whats-new-in-csharp-7-0/#user-content-ref-returns-and-locals). This lets NSubstitute have better support for working with `out` and `ref` arguments, but also means that test written using previous NSubstitute versions will now fail to compile with pre-C# 7 compilers.
+Argument matchers (`Arg.Is`, `Arg.Any` etc.) now use [`ref` returns which were introduced in C# 7.0](https://blogs.msdn.microsoft.com/dotnet/2016/08/24/whats-new-in-csharp-7-0/#user-content-ref-returns-and-locals). This lets NSubstitute have better support for working with `out` and `ref` arguments, but also means that test written using previous NSubstitute versions will now fail to compile with pre-C# 7 compilers with the following error:
+
+> CS7085: By-reference return type 'ref T' is not supported.
 
 Reason: Previous NSubstitute versions had quite limited support for `out` and `ref` arguments. Enhanced support for `out` and `ref` in C# 7 means we are likely to see these being used more frequently. As the vast majority of people using NSubstitute seem to be on C# 7+ (based on [NuGet statistics](https://www.nuget.org/stats/packages/NSubstitute?groupby=Version&groupby=ClientVersion)), we've thought it best to make sure NSubstitute's default behaviour works for these cases.
 
 Workaround: If at all possible please update to a recent version of your .NET compiler (C# 7+, VB 2017, VB 15.3+, F# 4.1+). This should provide support for `ref` returns and make the code compile fine with the new `Arg` methods. These shipped with Visual Studio 2017. Visual Studio for Mac 2017 and JetBrains Rider 2017 also support C# 7+.
 
 If it is not possible for you to use a C# 7-compatible compiler, we have added an `Arg.Compat` class which has all the same members as `Arg`, just without the `ref` return type. If you replace `Arg.` references with `Arg.Compat.` in your project then you can continue to use older compilers with NSubstitute 4.x. Alternatively you can use a `NSubstitute.Compatibility.CompatArg` instance in your fixture which may make migration a bit easier. Both these approaches are described in the [Compatibility argument matchers](http://nsubstitute.github.io/help/compat-args) documentation.
+
+---------------
+
+The argument matcher change to support `out` and `ref` mentioned above can also cause a compilation error if argument matchers are used in expression trees (see issue #471):
+
+> CS8153: An expression tree lambda may not contain a call to a method, property, or indexer that returns by reference
+
+Reason: `out` and `ref` support for argument matchers is important to address as it often caused confusion for people now that C# has more support for `out` and `ref`. We are always reluctant to introducing breaking changes, but in this case we estimated the benefit of the feature would outweigh the downside of no longer being able to use matchers in expression trees.
+
+Workaround:
+* Move the NSubstitute statement/assertion outside of the expression tree; or
+* Use an `Arg.Compat.` matcher as described above and in the [Compatibility argument matchers](http://nsubstitute.github.io/help/compat-args) documentation.
+
+For example:
+
+```
+// Given `void specify(Expression<Action> expectation)`, the following will fail to compile (CS8153):
+specify(() => sub.Received().SomeCall(Arg.Any<int>()));
+
+// Workaround 1: move out of expression tree
+sub.Received().SomeCall(Arg.Any<int>());
+// Workaround 2: use Arg.Compat
+specify(() => sub.Received().SomeCall(Arg.Compat.Any<int>()));
+```
 
 ---------------
 

--- a/docs/help/_posts/2013-04-15-compat-args.markdown
+++ b/docs/help/_posts/2013-04-15-compat-args.markdown
@@ -78,3 +78,39 @@ public class BaseTestFixture {
 {% endexamplecode %}
 
 If you are later able to update the C# compiler your project is using, you can remove the `CompatArg` field and all `Arg` references will go through standard arg matchers (and you'll now be able to use them with `out` and `ref` parameters!).
+
+## Argument matchers in expression trees
+
+As of NSubstitute 4.0 argument matchers can no longer be used in expression trees. Doing so will cause a compilation error:
+
+> CS8153: An expression tree lambda may not contain a call to a method, property, or indexer that returns by reference
+
+The `Arg.Compat` matchers can be used to work around this issue. They do not return by reference so they are fine to use in expression trees. 
+
+{% requiredcode %}
+// Pretending this takes an Expression<Action> arg (rather than adding a reference to
+// System.Linq.Expression).
+protected void specify(Action expectation) {
+    expectation();
+}
+{% endrequiredcode %}
+
+{% examplecode csharp %}
+public interface IFoo { void DoStuff(int i); }
+
+[Test]
+public void Sample() {
+    var sub = Substitute.For<IFoo>();
+    sub.DoStuff(42);
+    // If `specify` takes an `Expression<Action>` argument, this will fail with CS8153: 
+    //   specify(() => sub.Received().DoStuff(Arg.Any<int>()));
+
+    // Instead use compat matcher:
+    specify(() => sub.Received().DoStuff(Arg.Compat.Any<int>()));
+
+    // Or re-write without expression tree if possible:
+    sub.Received().DoStuff(Arg.Any<int>());
+}
+{% endexamplecode %}
+
+

